flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpointStore.java

```diff
@@ -60,6 +60,10 @@ default CompletedCheckpoint getLatestCheckpoint(boolean isPreferCheckpointForRec
 		}
 
 		CompletedCheckpoint candidate = getAllCheckpoints().get(getAllCheckpoints().size() - 1);
+		if (!candidate.getProperties().isSavepoint()) {
+			return candidate;
+		}
```

Comment:
This `if` check can be moved into the next `if` block. This would be more explicit that this check is specific only for `isPreferCheckpointForRecovery = true` and has no effect otherwise.

Additional suggestion: although it's not relevant to your PR and the changes in the ticket, you can assign a variable for `getAllCheckpoints()`  once in the top of the method. This would avoid additional method calls, array copies and should make this method more clearer.